VCFv4.3.tex

```diff
@@ -503,7 +506,41 @@ \subsubsection{Genotype fields}
   All phased genotypes that do not contain a PS subfield are assumed to belong to the same phased set.
   If the genotype in the GT field is unphased, the corresponding PS field is ignored.
   The recommended convention is to use the position of the first variant in the set as the PS identifier (although this is not required).
-\end{itemize}
+  \item LAA is a sorted list of n distinct integers, where 1 <= n <= $|ALT|$, giving the (1-based) indices within ALT of the alleles that are observed in the sample. 
+  In callsets with many samples, sites may grow to include numerous alternate alleles at the same POS. 
+  Usually, few of these alleles are actually observed in any one sample, but each genotype must supply fields like PL and AD for all of the alleles---a very inefficient representation as PL's size is quadratic in the allele count.
+  Similarly, in rare sites, which can be the bulk of the sites, the vast majority of the samples are reference. 
+  To prevent this growth in VCF size, one can choose to specify the genotype, allele depth and the genotype likelihood against a subset of ``Local Alleles".
+  LAA is the strictly increasing, 1-based index into ALT, pointing out the alternative alleles that are actually in-play for that sample. 
+  LAD is the depth of the local alleles, 
+  LPL is subset of the PL array that pertains to the alleles that are REF or referred to by LAA, 
+  LGT is the genotype but referencing the local alleles rather than the global ones.
+  It is implicit that REF is part of any ``local" context, and it always has index 0, even if the genotype is compound HET. 
+  For example, if REF is G, ALT is A,C,T,\verb!<*>! and a genotype only has information about G, C, and \verb!<*>!, one can have LAA=[2,4] and thus LPL will be interpreted as pertaining to the alleles [G, C, \verb!<*>!] and not contain likelihood values for genotypes that involve A or T. 
+  In this case LGT=0/1 means that the sample is G/C. 
+  GQ is still the genotype quality, even when the genotype is given against the local alleles. 
+  Note that reordering might be required and care need to be taken to reorder LAD and LPL appropriately.
+  LAA is required in order to interpret LAD, LPL, and LGT.
+  
+  In the following example, the records with the same POS encode the same information (some columns removed for clarity):
+ 
+  \begin{tabular}[l]{llllll}
+POS &REF&  ALT&FORMAT&sample\\
+1&G&A,C,T,\textless*\textgreater& LAA:LGT:LAD:LPL&  2,4:1/1:20,30,10:90,80,0,100,110,120\\
+1&G&A,C,T,\textless*\textgreater& GT:AD:PL&  2/2:20,.,30,.,10:90,.,.,80,.,0,.,.,.,.,100,.,110,.,120\\
+2&A&C,G,T,\textless*\textgreater& LAA:LGT:LAD:LPL& 3:0/1:15,25:40,0,80\\
+2&A&C,G,T,\textless*\textgreater& GT:AD:PL&0/3:15,.,.,25,.:40,.,.,.,.,.,0,.,.,80,.,.,.,.\\
+3&C&G,T,\textless*\textgreater& LAA:LGT:LAD:LPL& 4:0/0:30,1:0,30,80\\
+3&C&G,T,\textless*\textgreater& GT:AD:PL& 0/0:30,.,.1:0,.,.,.,.,.,.,.,.,.,30,.,.,.,80\\
+4&G&A,T,\textless*\textgreater& LAA:LGT:LAD:LPL& :0/0:30:0\\
```

Comment:
Typo here:  we either need to remove the LAA field and indicate that it can be left out if there are no non-ref local alleles, or we need to add a leading `.`

Reply:
I don't think this is a typo: the intention is that LAA is a 0-length list of integers (indicating that this record is homozygous ref for this sample and that none of the ALT alleles are in play), not an unknown MISSING value.

This may be a problem, as it's not clear that 0-length lists are valid syntax in VCF or BCF. Omitting LAA is probably not the right workaround, as presence of LAA is the simple indicator that local alleles are being used in a VCF record. Setting it to `.` (unknown) is somewhat misleading, as it **is** known…

Reply:
This example contradicts the current text, which says “LAA is a sorted list of _n_ distinct integers, where _n_ ≥ 1 […]”. So the text will need adjustment if this example record is deemed valid.

Reply:
`bcftools` converts the zero-length list into a missing value.  This comes from bcftools treating everything as BCF internally, and [BCF says zero-length lists are missing](https://samtools.github.io/hts-specs/VCFv4.4.pdf#subsubsection.6.3.3).

Possible fixes would be to say missing `LAA` means REF; to explicitly put 0 in `LAA` for REF-only; or use `GT` in these rows, but allow use of `LPL`, `LAD` etc. in the specific case that `GT` is REF-only.

Reply:
It makes sense to use the missing value `.` and state it explicitly that it means REF. I can't think of a case where this could cause problems.

Reply:
I agree with Petr here.  While the BCF specification states that a zero length array is defined to be a MISSING array, hence `.`, we could add an explicit statement to LAA saying that missing LAAs means that it matches REF.  What we gain is a solution to this problem.  What we lose is the ability to store an unknown LAA list.  I'm unsure what the latter practically means.  Why would you need to explicitly state it's missing instead of just omitting it.  If it's missing then we're not using LAA (and if the PL, AD etc are missing we're not defining it that way either).

PS. Technically even if it's `.` this still contradicts "LAA is a sorted list of n distinct integers" as `.` isn't an integer.  However that's nit-picking and maybe somewhere else VCF already defines `.` as being an integer of unknown value?  I don't know what we do elsewhere when we have this technical disparty.  Is it just assumed that "." is a polymophic type that matches whatever the item wants it to be?

Reply:
What the BCF spec says is that a type descriptor byte with ‘size’=0 denotes MISSING. It does not say what a type descriptor byte with ‘size’=15 followed by a typed Integer with value 0 denotes, but that would be a natural representation of a 0-length list. (The spec does say that ‘size’=15 implies the list length is ≥15, but the typed Integer is not biased so no meanings are described for typed Integer values <15.) Thus BCF does not provide a binding answer on whether 0-length lists are valid or representable in VCF/BCF.

But I'm not going to advocate blessing 0-length lists in VCF or BCF, especially not just for this, even though SAM found them useful for regularity (cf #135).

IMHO the clearest way to represent this corner case would be

> As a special case, the homozygous REF case in which no ALT alleles are observed in a particular sample is represented as LAA=[0].

(And we'd need to rework _n_ in the length maths: e.g. by dropping _n_ from the description of LAA, and adding “where _n_ is the number of **ALT** alleles listed in LAA” to the others.)

Reply:
To expand on my previous comment: the logic of LAA is to define a list of alleles that can be referred to in subsequent localized fields. If only the reference allele is relevant, then it makes sense to provide the missing value `.`, in this sense it behaves in exactly the same way as what we write in REF and ALT columns. Note also the naming of the tag stands for "Local Alternate Alleles". By using `.` we are not introducing a new concept.

Reply:
Parachuting in here a bit, so my apologies if I'm missing context, but defining LAA as an R-length array whose first element is always the number zero both avoids these nasal demons and avoids a branch [1] when connecting a local index into a global index.

Is it too late to consider an R-length LAA instead of an A-length one?


[1] Text processing cost probably dominates, but if you convert a VCF into some binary format without adjusting LAA, maybe this matters a bit.